src/value/ValueTransfer.java

```diff
@@ -0,0 +1,235 @@
+                    resultRange = leftRange.shiftLeft(rightRange);
+                    break;
+                case SIGNED_SHIFT_RIGHT:
+                    resultRange = leftRange.signedShiftRight(rightRange);
+                    break;
+                case UNSIGNED_SHIFT_RIGHT:
+                    resultRange = leftRange.unsignedShiftRight(rightRange);
+                    break;
+                case BITWISE_AND:
+                    resultRange = leftRange.bitwiseAnd(rightRange);
+                    break;
+                case BITWISE_OR:
+                    resultRange = leftRange.bitwiseOr(rightRange);
+                    break;
+                case BITWISE_XOR:
+                    resultRange = leftRange.bitwiseXor(rightRange);
+                    break;
+                default:
+                    throw new BugInCF("ValueTransfer: unsupported operation: " + op);
+            }
+            // Any integral type with less than 32 bits would be promoted to 32-bit int type during
+            // operations.
+            return leftNode.getType().getKind() == TypeKind.LONG
+                || rightNode.getType().getKind() == TypeKind.LONG
+                ? resultRange
+                : resultRange.intRange();
+        } else {
+            return Range.EVERYTHING;
+        }
+    }
+
+    /** Get possible integer range from annotation. */
+    private Range getIntRange(Node subNode, TransferInput<CFValue, CFStore> p) {
+        AnnotationMirror val = getValueAnnotation(subNode, p);
+        return getIntRangeFromAnnotation(subNode, val);
+    }
+
+    /**
+     * Returns the {@link Range} object corresponding to the annotation {@code val} casted to the
+     * type of {@code node}.
+     *
+     * @param node a node
+     * @param val annotation mirror
+     * @return the {@link Range} object corresponding to the annotation {@code val} casted to the
+     *     type of {@code node}.
+     */
+    private Range getIntRangeFromAnnotation(Node node, AnnotationMirror val) {
+        Range range;
+        if (val == null
+            || AnnotationUtils.areSameByName(val, typeFactory.UNKNOWNVAL)) {
+            range = Range.EVERYTHING;
+        } else if (typeFactory.isIntRange(val)) {
+            range = ValueAnnotatedTypeFactory.getRange(val);
+        } else if (AnnotationUtils.areSameByName(val, typeFactory.BOTTOMVAL)) {
+            return Range.NOTHING;
+        } else {
+            range = Range.EVERYTHING;
+        }
+        return NumberUtils.castRange(node.getType(), range);
+    }
+
+    /**
+     * Returns true if {@code node} an integral type and is {@code anno} is {@code @UnknownVal}.
+     *
+     * @param node a node
+     * @param anno annotation mirror
+     * @return true if node is annotated with {@code @UnknownVal} and it is an integral type.
+     */
+    private boolean isIntegralUnknownVal(Node node, AnnotationMirror anno) {
+        return AnnotationUtils.areSameByName(anno, typeFactory.UNKNOWNVAL)
```

Comment:
Once you are in `ValueInferenceTransfer`, I believe this anno will be a `VarAnno` and you won't be able to do this check. You will need to get the constant slot from the `VarAnno` and then the anno from constant slot.
Something like this:
```
Slot s = slotManager.getVariableSlot(anno);
if s == ConstantSlot then s.getValue();
```
Then you can do this check against `s`.